Added readDeepNullable and similar methods

jroper opened the pull request:
Fixes #268

This variant of readNullable successfully reads deep paths even when nodes before the last are not present.

An alternative to this change might be to simply change the behavior of `readNullable`. It has been this way since https://github.com/playframework/playframework/pull/662, but there is no justification for why anywhere that I can find. Typically, `readNullable` is only used to read a path that is one level deep, so I don't imagine changing its behavior for deeper paths to be a big issue. I can't imagine any use case where you wouldn't `None` when nodes before the last path are not present.

lforite commented:
@mandubian any input on this ?

renatocaval commented on play-json/shared/src/main/scala/play/api/libs/json/JsConstraints.scala:
```diff
@@ -71,6 +76,42 @@ trait PathReads {
     }
 
+  /**
+   * Reads a Option[T] search optional or nullable field at JsPath (field not found or null is None
```
```suggestion
   * Reads an Option[T] search optional or nullable field at JsPath (field not found or null is None
```

renatocaval commented on play-json/shared/src/main/scala/play/api/libs/json/JsConstraints.scala:
```diff
@@ -71,6 +76,42 @@ trait PathReads {
+    deepNullableWithDefault(path, None)
+
+  /**
+   * Reads an Option[T] search optional or nullable field at JsPath (field not found replaced by
```
```suggestion
   * Reads an Option[T] (optional or nullable field) at JsPath (field not found replaced by
```

renatocaval commented on play-json/shared/src/main/scala/play/api/libs/json/JsPath.scala:
```diff
@@ -311,6 +311,34 @@ case class JsPath(path: List[PathNode] = List()) {
   def readNullableWithDefault[T](defaultValue: => Option[T])(implicit r: Reads[T]): Reads[Option[T]] = Reads.nullableWithDefault[T](this, defaultValue)(r)
 
+  /**
+   * Reads a Option[T] search optional or nullable field at JsPath (field not found or null is None
```
```suggestion
   * Reads an Option[T] (optional or nullable field) at JsPath (field not found or null is None
```

renatocaval commented on play-json/shared/src/main/scala/play/api/libs/json/JsPath.scala:
```diff
@@ -439,6 +467,24 @@ case class JsPath(path: List[PathNode] = List()) {
   }
 
+  /**
+   * Reads/Writes a Option[T] (optional or nullable field) at given JsPath
```
```suggestion
   * Reads/Writes an Option[T] (optional or nullable field) at given JsPath
```

renatocaval commented on play-json/shared/src/main/scala/play/api/libs/json/JsPath.scala:
```diff
@@ -311,6 +311,34 @@ case class JsPath(path: List[PathNode] = List()) {
+  def readDeepNullable[T](implicit r: Reads[T]): Reads[Option[T]] = Reads.deepNullable[T](this)(r)
+
+  /**
+   * Reads an Option[T] search optional or nullable field at JsPath (field not found replaced by
```
```suggestion
   * Reads an Option[T] (optional or nullable field) at JsPath (field not found replaced by
```

renatocaval commented on play-json/shared/src/main/scala/play/api/libs/json/JsPath.scala:
```diff
@@ -439,6 +467,24 @@ case class JsPath(path: List[PathNode] = List()) {
+  def formatDeepNullable[T](implicit f: Format[T]): OFormat[Option[T]] = Format.deepNullable[T](this)(f)
+
+  /**
+   * Reads/Writes a Option[T] (nullable field) at given JsPath
```
```suggestion
   * Reads/Writes an Option[T] (nullable field) at given JsPath
```

cchantep commented on play-json/shared/src/main/scala/play/api/libs/json/JsConstraints.scala:
```diff
@@ -71,6 +76,42 @@ trait PathReads {
+   * - If the last node in JsPath is found with value "null" => returns None
+   * - If the entire path is found => applies implicit Reads[T]
+   */
+  def deepNullableWithDefault[A](path: JsPath, defaultValue: => Option[A])(implicit reads: Reads[A]): Reads[Option[A]] =
```
Shouldn't it been the default behaviour for the existing nullable `Reads` ?

renatocaval replied:
Yes, you are right. I'm ok to not introduce new methods and simply 'fix' the behaviour of the existing ones. 

@jroper, do you want to adapt your PR accordingly?

jroper commented:
@renatocaval I've updated to change the existing `readNullable` instead of introducing a new method.

Since I was changing such a central method, I wanted to ensure that I wasn't introducing a performance regression, so I created a new JMH benchmark that reads a case class with 3 optional fields from a JsValue that has two of them set. There was a significant performance increase, from 2650476 ops/s to 4155399 ops/s, a 56% performance improvement (note, this test is designed to execute that exact code and nothing else, so it's not going to have anywhere near as big an impact in the real world, but still its significant).

So realising that this code was really performance sensitive, and seeing that the code that I replaced it with, in the most common case, did a lot of allocations and unnecessary work, I decided to implement a fast path in  in `JsPath.asSingleJson` specifically targeted at `(__ \ "foo").readNullable[Foo]`, that is, reading from a path with a single key in it, which is by far the most common use of `readNullable`. Performance went up to 7569083 ops/second, a further 82% increase, for a total of 185% performance increase on the old implementation. Not bad!

Then I realised the same could be done for `(__ \ "foo").read[Foo]`, which uses `JsPath.asSingleJsonResult`. I didn't write a new test for this one, instead I reused the existing deserialise list one, which exercises a much broader set of code so we would expect the optimisation to have less of an impact (but be more meaningful), it went from 11664 ops/second to 16009 ops/second, a 37% performance increase.
